### PR TITLE
Rewrite REAME.md for DS2 and update examples.

### LINKED ISSUE #245: Completely rewrite README.md for DS2.

### PR DESCRIPTION
**Pull request by xinghai-sun**
Resolve #245 and #235

### CODE REVIEW

File: deep_speech_2/test.py [hunk 1]
```diff
@@ -36,14 +36,14 @@
...
-        'data/librispeech/eng_vocab.txt',
+        'data/librispeech/vocab.txt',
         "Filepath of vocabulary.")
 add_arg('model_path',       str,
         './checkpoints/params.latest.tar.gz',
         "If None, the training starts from scratch, "
         "otherwise, it resumes from the pre-trained model.")
 add_arg('lang_model_path',  str,
-        'lm/data/common_crawl_00.prune01111.trie.klm',
+        'model_zoo/lm/common_crawl_00.prune01111.trie.klm',
```
Comment (kuke) [inline deep_speech_2/test.py]: ```model_zoo```->```models```?

File: deep_speech_2/tools/tune.py [hunk 2]
```diff
@@ -41,10 +41,10 @@
         'data/librispeech/mean_std.npz',
         "Filepath of normalizer's mean & std.")
 add_arg('vocab_path',       str,
-        'data/librispeech/eng_vocab.txt',
+        'data/librispeech/vocab.txt',
         "Filepath of vocabulary.")
 add_arg('lang_model_path',  str,
-        'lm/data/common_crawl_00.prune01111.trie.klm',
+        'model_zoo/lm/common_crawl_00.prune01111.trie.klm',
```
Comment (kuke) [inline deep_speech_2/tools/tune.py]: ```models```

File: deep_speech_2/infer.py [hunk 3]
```diff
@@ -35,10 +35,10 @@
         'data/librispeech/mean_std.npz',
         "Filepath of normalizer's mean & std.")
 add_arg('vocab_path',       str,
-        'data/librispeech/eng_vocab.txt',
+        'data/librispeech/vocab.txt',
         "Filepath of vocabulary.")
 add_arg('lang_model_path',  str,
-        'lm/data/common_crawl_00.prune01111.trie.klm',
+        'model_zoo/lm/common_crawl_00.prune01111.trie.klm',
```
Comment (kuke) [inline deep_speech_2/infer.py]: models

File: deep_speech_2/examples/librispeech/run_infer.sh [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+--alpha=0.36 \
+--beta=0.25 \
+--cutoff_prob=0.99 \
+--use_gru=False \
+--use_gpu=True \
+--share_rnn_weights=True \
+--infer_manifest='data/librispeech/manifest.test-clean' \
+--mean_std_path='data/librispeech/mean_std.npz' \
+--vocab_path='data/librispeech/vocab.txt' \
+--model_path='checkpoints/libri/params.latest.tar.gz' \
```
Comment (kuke) [inline deep_speech_2/examples/librispeech/run_infer.sh:33]: In some other files, this path is  ```'checkpoints/librispeech/*```. Please unify them.

File: deep_speech_2/README.md [hunk 5]
```diff
@@ -1,146 +1,393 @@
...
 ```
 
-Please replace `$PADDLE_INSTALL_DIR` with your own paddle installation directory.
+## Getting Started
 
-## Usage
+Several shell scripts provided in `./examples` will help us to quickly give it a try, for most major modules, including data preparation, model training, case inference and model evaluation, with a few public dataset (e.g. [LibriSpeech](http://www.openslr.org/12/), [Aishell](https://github.com/kaldi-asr/kaldi/tree/master/egs/aishell)). Reading these examples will also help us understand how to make it work with our own data.
 
-### Preparing Data
+Some of the scripts in `./examples` are configured with 8 GPUs. If you don't have 8 GPUs available, please modify `CUDA_VISIBLE_DEVICE` and `--trainer_count`. If you don't have any GPU available, please set `--use_gpu` to False to use CPUs instead.
```
Comment (kuke) [inline deep_speech_2/README.md]: ```CUDA_VISIBLE_DEVICE``` -> ```CUDA_VISIBLE_DEVICES```

File: deep_speech_2/README.md [hunk 6]
```diff
@@ -1,146 +1,393 @@
...
+- [Hyper-parameters Tuning](#hyper-parameters-tuning)
+- [Training for Mandarin Language](#training-for-mandarin-language)
+- [Trying Live Demo with Your Own Voice](#trying-live-demo-with-your-own-voice)
+- [Experiments and Benchmarks](#experiments-and-benchmarks)
+- [Released Models](#released-models)
+- [Questions and Help](#questions-and-help)
+
+## Prerequisites
+- Only support Python 2.7
+- PaddlePaddle the latest version (please refer to the [Installation Guide](https://github.com/PaddlePaddle/Paddle#installation))
```
Comment (luotao1) [inline deep_speech_2/README.md]: PaddlePaddle: the latest version

**Review by luotao1 — COMMENTED** [review]
- Readme.md文档逻辑很清晰，但英文语法上面存在挺多小问题，需要之后再refine一下。
- 另外，其他models下的readme都为中文，deep speech2也放一个中文的介绍吧，有助于帮助国内用户尽快上手，且方便大家贡献。

File: deep_speech_2/README.md [hunk 7]
```diff
@@ -1,146 +1,393 @@
 # DeepSpeech2 on PaddlePaddle
 
->TODO: to be updated, since the directory hierarchy was changed.
+*DeepSpeech2 on PaddlePaddle* is an open-source implementation of end-to-end Automatic Speech Recognition (ASR) engine, based on [Baidu's Deep Speech 2 paper](http://proceedings.mlr.press/v48/amodei16.pdf), with [PaddlePaddle](https://github.com/PaddlePaddle/Paddle) platform. Our vision is to empower both industrial application and academic research on speech recognition, via an easy-to-use, efficient and scalable implementation, including training, inferencing & testing module, distributed [PaddleCloud](https://github.com/PaddlePaddle/cloud) training, and demo deployment. Besides, several pre-trained models for both English and Mandarin are also released.
```
Comment (pkuyym) [inline deep_speech_2/README.md]: training, inferencing --> training, inference?

File: deep_speech_2/README.md [hunk 8]
```diff
@@ -1,146 +1,393 @@
...
+- [Released Models](#released-models)
+- [Questions and Help](#questions-and-help)
+
+## Prerequisites
+- Only support Python 2.7
+- PaddlePaddle the latest version (please refer to the [Installation Guide](https://github.com/PaddlePaddle/Paddle#installation))
 
 ## Installation
 
+Please install the [prerequisites](#prerequisites) above before moving on.
```
Comment (pkuyym) [inline deep_speech_2/README.md]: --> Please make sure above prerequisites has been satisfied before moving on.
```install prerequisites``` may be not proper.

File: deep_speech_2/README.md [hunk 9]
```diff
@@ -1,146 +1,393 @@
...
+git clone https://github.com/PaddlePaddle/models.git
+cd models/deep_speech_2
 sh setup.sh
 ```
 
-Please replace `$PADDLE_INSTALL_DIR` with your own paddle installation directory.
+## Getting Started
 
-## Usage
+Several shell scripts provided in `./examples` will help us to quickly give it a try, for most major modules, including data preparation, model training, case inference and model evaluation, with a few public dataset (e.g. [LibriSpeech](http://www.openslr.org/12/), [Aishell](https://github.com/kaldi-asr/kaldi/tree/master/egs/aishell)). Reading these examples will also help us understand how to make it work with our own data.
```
Comment (pkuyym) [inline deep_speech_2/README.md]: us --> you
https://github.com/kaldi-asr/kaldi/tree/master/egs/aishell --> http://www.openslr.org/33
our --> your

File: deep_speech_2/README.md [hunk 10]
```diff
@@ -1,146 +1,393 @@
...
+    ```
+
+    Notice that this is only a toy example with a tiny sampled subset of LibriSpeech. If we would like to try with the complete dataset (would take several days for training), please go to `examples/librispeech` instead.
+- Prepare the data
+
+    ```
+    sh run_data.sh
+    ```
+
+    `run_data.sh` will download dataset, generate manifests, collect normalizer' statistics and build vocabulary. Once the data preparation is done, we will find the data (only part of LibriSpeech) downloaded in `~/.cache/paddle/dataset/speech/libri` and the corresponding manifest files generated in `./data/tiny` as well as a mean stddev file and a vocabulary file. It has to be run for the very first time we run this dataset and is reusable for all further experiments.
```
Comment (pkuyym) [inline deep_speech_2/README.md]: normalizer'  --> normalizer
we --> you

File: deep_speech_2/README.md [hunk 11]
```diff
@@ -1,146 +1,393 @@
...
+- [Distributed Cloud Training](#distributed-cloud-training)
+- [Hyper-parameters Tuning](#hyper-parameters-tuning)
+- [Training for Mandarin Language](#training-for-mandarin-language)
+- [Trying Live Demo with Your Own Voice](#trying-live-demo-with-your-own-voice)
+- [Experiments and Benchmarks](#experiments-and-benchmarks)
+- [Released Models](#released-models)
+- [Questions and Help](#questions-and-help)
+
+## Prerequisites
+- Only support Python 2.7
```
Comment (kuke) [inline deep_speech_2/README.md]: Python 2.7 only supported

File: deep_speech_2/README.md [hunk 12]
```diff
@@ -1,146 +1,393 @@
...
 ```
-python datasets/librispeech/librispeech.py --help
+{"audio_filepath": "/home/work/.cache/paddle/Libri/134686/1089-134686-0001.flac", "duration": 3.275, "text": "stuff it into you his belly counselled him"}
+{"audio_filepath": "/home/work/.cache/paddle/Libri/134686/1089-134686-0007.flac", "duration": 4.275, "text": "a cold lucid indifference reigned in his soul"}
 ```
 
-### Preparing for Training
+To use your custom data, you only need to generate such manifest files to summarize the dataset. Given such summarized manifests, training, inference and all other modules can be aware of where to access the audio files, as well as their meta data including the transcription labels.
+
+For how to generate such manifest files, please refer to `data/librispeech/librispeech.py`, which download and generate manifests for LibriSpeech dataset.
```
Comment (kuke) [inline deep_speech_2/README.md]: ```downloads``` & ```generates```

File: deep_speech_2/README.md [hunk 13]
```diff
@@ -1,146 +1,393 @@
...
+    ```
+
+    `run_data.sh` will download dataset, generate manifests, collect normalizer' statistics and build vocabulary. Once the data preparation is done, we will find the data (only part of LibriSpeech) downloaded in `~/.cache/paddle/dataset/speech/libri` and the corresponding manifest files generated in `./data/tiny` as well as a mean stddev file and a vocabulary file. It has to be run for the very first time we run this dataset and is reusable for all further experiments.
+- Train your own ASR model
+
+    ```
+    sh run_train.sh
+    ```
+
+    `run_train.sh` will start a training job, with training logs printed to stdout and model checkpoint of every pass/epoch saved to `./checkpoints/tiny`. We can resume the training from these checkpoints, or use them for inference, evaluation and deployment.
```
Comment (pkuyym) [inline deep_speech_2/README.md]: ```resume the training``` is not rigorous.

File: deep_speech_2/README.md [hunk 14]
```diff
@@ -1,146 +1,393 @@
...
+- Case inference with an existing model
+
+    ```
+    sh run_infer.sh
+    ```
+
+    `run_infer.sh` will show us some speech-to-text decoding results for several (default: 10) samples with the trained model. The performance might not be good now as the current model is only trained with a toy subset of LibriSpeech. To see the results with a better model, we can download a well-trained (trained for several days, with the complete LibriSpeech) model and do the inference:
```
Comment (pkuyym) [inline deep_speech_2/README.md]: we --> you

File: deep_speech_2/README.md (hunk 12 above)
Comment (pkuyym) [inline deep_speech_2/README.md]: download --> downloads
generate --> generates
manifests --> manifest files

File: deep_speech_2/README.md [hunk 15]
```diff
@@ -1,146 +1,393 @@
...
+    --init_model_path CHECKPOINT_PATH_TO_RESUME_FROM
+    ```
 
-More help for arguments:
+For more help on arguments:
 
 ```
 python train.py --help
 ```
+or refer to `example/librispeech/run_train.sh`.
```
Comment (pkuyym) [inline deep_speech_2/README.md]: run_train.sh doesn't include all arguments like ```init_model_path```

File: deep_speech_2/README.md [hunk 16]
```diff
@@ -1,146 +1,393 @@
...
-The following steps, inference, parameters tuning and evaluating, will require a language model during decoding.
-A compressed language model is provided and can be accessed by
+Data augmentation has often been a highly effective technique to boost the deep learning performance. We augment our speech data by synthesizing new audios with small random perturbation (label-invariant transformation) added upon raw audios. We don't have to do the syntheses by ourselves, as it is already embedded into the data provider and is done on the fly, randomly for each epoch during training.
 
-```
-cd ./lm
-sh run.sh
-cd ..
-```
+Six optional augmentation components are provided for us to configured and inserted into the processing pipeline.
```
Comment (pkuyym) [inline deep_speech_2/README.md]: Six optional augmentation components are provided which can be configured and inserted into the processing pipeline

**Review by kuke — COMMENTED** [review]
Some minor problems, almost LGTM

File: deep_speech_2/README.md [hunk 17]
```diff
@@ -149,22 +396,37 @@ brew install portaudio
...
 ```
-CUDA_VISIBLE_DEVICES=0 python demo_server.py
+CUDA_VISIBLE_DEVICES=0 \
+python -u deploy/demo_client.py \
+--host_ip 'localhost' \
+--host_port 8086
 ```
-And then in another console, start the demo's client:
+
+Now, in the client console, press the `whitespace` key, hold, and start speaking. Until we finish our utterance, we release the key to let the speech-to-text results shown in the console. To quit the client, just press `ESC` key.
```
Comment (pkuyym) [inline deep_speech_2/README.md]: Here, we can paste the information message after starting client console.

File: deep_speech_2/README.md [hunk 18]
```diff
@@ -149,22 +396,37 @@ brew install portaudio
...
+
+Notice that `deploy/demo_client.py` must be run in a machine with a microphone device, while `deploy/demo_server.py` could be run in one without any audio recording hardware, e.g. any remote server machine. Just be careful to set the `host_ip` and `host_port` argument with the actual accessible IP address and port, if the server and client are running with two separate machines. Nothing should be done if they are running in one single machine.
```
Comment (pkuyym) [inline deep_speech_2/README.md]: in --> on

File: deep_speech_2/README.md (hunk 6 above)
Comment (xinghai-sun) [reply deep_speech_2/README.md]: 这是临时写的。由于目前正式发布的版本都不适用ds2，无法写版本号，有什么建议吗？写commit hash可能更不合适。
此外，我们可以额外提供一个包含ds依赖的 docker image。

File: deep_speech_2/test.py (hunk 1 above)
Comment (xinghai-sun) [reply deep_speech_2/test.py]: Done.

File: deep_speech_2/README.md (hunk 7 above)
Comment (xinghai-sun) [reply deep_speech_2/README.md]: Done.

File: deep_speech_2/README.md (hunk 11 above)
Comment (xinghai-sun) [reply deep_speech_2/README.md]: Done.

File: deep_speech_2/README.md (hunk 8 above)
Comment (xinghai-sun) [reply deep_speech_2/README.md]: Done.

File: deep_speech_2/README.md (hunk 9 above)
Comment (xinghai-sun) [reply deep_speech_2/README.md]: Done.

File: deep_speech_2/README.md (hunk 5 above)
Comment (xinghai-sun) [reply deep_speech_2/README.md]: Done.

File: deep_speech_2/README.md (hunk 10 above)
Comment (xinghai-sun) [reply deep_speech_2/README.md]: normalizer's
done.

File: deep_speech_2/README.md (hunk 13 above)
Comment (xinghai-sun) [reply deep_speech_2/README.md]: "resume training" is ok, e.g. https://cn.mathworks.com/help/nnet/ug/resume-training-from-a-checkpoint-network.html

File: deep_speech_2/README.md (hunk 14 above)
Comment (xinghai-sun) [reply deep_speech_2/README.md]: Done.

File: deep_speech_2/README.md (hunk 12 above)
Comment (xinghai-sun) [reply deep_speech_2/README.md]: --> will download and generate

Comment (xinghai-sun) [reply deep_speech_2/README.md]: Done.

File: deep_speech_2/README.md (hunk 16 above)
Comment (xinghai-sun) [reply deep_speech_2/README.md]: --> Six optional augmentation components are provided to be selected, configured and inserted into the processing pipeline.

File: deep_speech_2/README.md (hunk 18 above)
Comment (xinghai-sun) [reply deep_speech_2/README.md]: Done.

File: deep_speech_2/examples/librispeech/run_infer.sh (hunk 4 above)
Comment (xinghai-sun) [reply deep_speech_2/examples/librispeech/run_infer.sh:33]: Done. Unified with 'checkpoints/libri/*'

File: deep_speech_2/infer.py (hunk 3 above)
Comment (xinghai-sun) [reply deep_speech_2/infer.py]: Done.

File: deep_speech_2/tools/tune.py (hunk 2 above)
Comment (xinghai-sun) [reply deep_speech_2/tools/tune.py]: Done.

**Review by kuke — COMMENTED** [review]
LGTM

File: deep_speech_2/data/librispeech/librispeech.py [hunk 19]
```diff
@@ -60,14 +58,13 @@
...
-    """
-    Download file from url to target_dir, and check md5sum.
+    """Download file from url to target_dir, and check md5sum.
     """
     if not os.path.exists(target_dir): os.makedirs(target_dir)
     filepath = os.path.join(target_dir, url.split("/")[-1])
     if not (os.path.exists(filepath) and md5file(filepath) == md5sum):
         print("Downloading %s ..." % url)
-        os.system("wget -c " + url + " -P " + target_dir)
+        ret = os.system("wget -c " + url + " -P " + target_dir)
```
Comment (pkuyym) [inline deep_speech_2/data/librispeech/librispeech.py]: remove ```ret```

File: deep_speech_2/README.md [hunk 20]
```diff
@@ -1,170 +1,432 @@
...
 ```
 
-Please replace `$PADDLE_INSTALL_DIR` with your own paddle installation directory.
+## Getting Started
 
-## Usage
+Several shell scripts provided in `./examples` will help us to quickly give it a try, for most major modules, including data preparation, model training, case inference and model evaluation, with a few public dataset (e.g. [LibriSpeech](http://www.openslr.org/12/), [Aishell](http://www.openslr.org/33)). Reading these examples will also help you to understand how to make it work with your own data.
 
-### Preparing Data
+Some of the scripts in `./examples` are configured with 8 GPUs. If you don't have 8 GPUs available, please modify `CUDA_VISIBLE_DEVICES` and `--trainer_count`. If you don't have any GPU available, please set `--use_gpu` to False to use CPUs instead.
```
Comment (pkuyym) [inline deep_speech_2/README.md]: Prompt OOM problem.

File: deep_speech_2/data/librispeech/librispeech.py (hunk 19 above)
Comment (xinghai-sun) [reply deep_speech_2/data/librispeech/librispeech.py]: Done.

File: deep_speech_2/README.md (hunk 20 above)
Comment (xinghai-sun) [reply deep_speech_2/README.md]: Done.

**Review by pkuyym — APPROVED** [review]
LGTM
